src/Arcus.Scripting.ApiManagement/Scripts/Remove-AzApiManagementDefaults.ps1

```diff
@@ -5,32 +5,71 @@ param(
-$apiResult = Remove-AzApiManagementApi -Context $apimContext -ApiId 'echo-api'
-
-Write-Host "Removing Starter product..."
-$starterResult = Remove-AzApiManagementProduct -Context $apimContext -ProductId 'starter' -DeleteSubscriptions
-
-Write-Host "Removing Unlimited product..."
-$unlimitedResult = Remove-AzApiManagementProduct -Context $apimContext -ProductId 'unlimited' -DeleteSubscriptions
+Write-Host "Checking if 'echo' API exists..."
+$echoExists = $true
+try {
+    $apiGetResult = Get-AzApiManagementApi -Context $apimContext -ApiId 'echo-api' -ErrorAction Stop
+}
+catch {
+    $echoExists = $false
+    Write-Host "The 'echo' API does not exist, skipping removal..."
+}
+if ($echoExists) {
+    try {
+        Write-Host "Removing 'echo' API..."
+        $apiRemoveResult = Remove-AzApiManagementApi -Context $apimContext -ApiId 'echo-api' -ErrorAction Stop
+    }
+    catch {
+        Write-Error "Failed to remove the 'echo' API"
+        $exceptionOccurredOnDelete = $true
+    }
+}
 
-$message = $null
+Write-Host "Checking if 'starter' product exists..."
+$starterExists = $true
+try {
+    $starterGetResult = Get-AzApiManagementProduct -Context $apimContext -ProductId 'starter' -ErrorAction Stop
+}
+catch {
+    $starterExists = $false
+    Write-Host "The 'starter' product does not exist, skipping removal..."
+}
+if ($starterExists) { 
+    try {
+        Write-Host "Removing 'starter' product..."
+        $starterRemoveResult = Remove-AzApiManagementProduct -Context $apimContext -ProductId 'starter' -DeleteSubscriptions -ErrorAction Stop
+    }
+    catch {
+        Write-Error "Failed to remove the 'starter' product"
+        $exceptionOccurredOnDelete = $true
+    }
+}
 
-if ($null -ne $apiResult) {
-    $message += [System.Environment]::NewLine + "> Failed to remove the 'echo' API"
+Write-Host "Checking if 'unlimited' product exists..."
+$unlimitedExists = $true
+try {
+    $unlimitedGetResult = Get-AzApiManagementProduct -Context $apimContext -ProductId 'unlimited' -ErrorAction Stop
 }
-if ($null -ne $starterResult) {
-    $message += [System.Environment]::NewLine + "> Failed to remove the 'starter' Product"
+catch {
+    $unlimitedExists = $false
+    Write-Host "The 'unlimited' product does not exist, skipping removal..."
 }
-if ($null -ne $unlimitedResult) {
-    $message += [System.Environment]::NewLine + "> Failed to remove the 'unlimited' Product"
+if ($unlimitedExists) { 
+    try {
+        Write-Host "Removing 'unlimited' product..."
+        $unlimitedRemoveResult = Remove-AzApiManagementProduct -Context $apimContext -ProductId 'unlimited' -DeleteSubscriptions -ErrorAction Stop
+    }
+    catch {
+        Write-Error "Failed to remove the 'unlimited' product"
+        $exceptionOccurredOnDelete = $true
+    }
 }
 
-if($null -eq $message){
-    Write-Host "Successfully removed the 'echo-api' API, 'starter' Product and 'unlimited' Product"
-}else{
-    Write-Error $message
+if ($exceptionOccurredOnDelete)
+{
+    throw
```

Comment:
Maybe we can keep track of which API's failed to remove and list them here in the exception message? Not sure what exception message will show when we only use `throw`. Would it use the previously written messages to the error stream?

Reply:
I improved the error handling, the catch for the GET actions now check if it is actually a `NotFound` response and handle it accordingly.

The final throw now also includes an error message which contains the steps that failed, for example:
`These action(s) failed: getting the 'echo-api', getting the 'starter' product`

See [`2162678` (#254)](https://github.com/arcus-azure/arcus.scripting/pull/254/commits/2162678fac83c2b57cbd2262d2f68ebf2772eb3d)